cmd/kubeadm/app/phases/certs/certlist.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certs
+
+import (
+	"crypto/x509"
+
+	certutil "k8s.io/client-go/util/cert"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/certs/pkiutil"
+)
+
+type configMutatorsFunc func(*kubeadmapi.InitConfiguration, *certutil.Config) error
+
+// KubeadmCert represents a certificate that Kubeadm will create to function properly.
+type KubeadmCert struct {
+	Name     string
+	BaseName string
+	CAName   string
+	// Some attributes will depend on the InitConfiguration, only known at runtime.
+	// These functions will be run in series, passed both the InitConfiguration and a cert Config.
+	configMutators []configMutatorsFunc
+	config         certutil.Config
+}
+
+// GetConfig returns the definition for the given cert given the provided InitConfiguration
+func (k *KubeadmCert) GetConfig(cfg *kubeadmapi.InitConfiguration) (*certutil.Config, error) {
+	for _, f := range k.configMutators {
+		if err := f(cfg, &k.config); err != nil {
+			return nil, err
+		}
+	}
+
+	return &k.config, nil
+}
+
+func init() {
+	CertMap = make(map[string]*KubeadmCert)
+	// initialize CertMap
+	for _, cert := range CertList {
+		CertMap[cert.Name] = cert
+	}
+}
+
+// CertMap is all the certificates in CertList, indexed by Name.
+var CertMap map[string]*KubeadmCert
+
+// CertList is all of the certificates kubeadm requires to function.
+var CertList = []*KubeadmCert{
```

Comment:
Instead of globals what do you think about defining a type of something like, `type Certificates []*KubeadmCert` and then this particular set of certificates can be behind a function like `func DefaultKubeadmCertificates() Certificates {...}` and then it's also accessible as a map `func (c Certificates) AsMap() map[string]*KubeadmCert`?

This will remove the need for global vars as well as the init() function. Generally I find init functions and global variables hard to work with.

Reply:
This type could also have the makeCertTree function on it